src/compiler/interpreter.jl

```diff
@@ -0,0 +1,110 @@
+        )
+    end
+end
+
+Core.Compiler.InferenceParams(interp::EnzymeInterpeter) = interp.inf_params
+Core.Compiler.OptimizationParams(interp::EnzymeInterpeter) = interp.opt_params
+Core.Compiler.get_world_counter(interp::EnzymeInterpeter) = interp.world
+Core.Compiler.get_inference_cache(interp::EnzymeInterpeter) = interp.local_cache
+Core.Compiler.code_cache(interp::EnzymeInterpeter) = WorldView(interp.global_cache, interp.world)
+
+# No need to do any locking since we're not putting our results into the runtime cache
+Core.Compiler.lock_mi_inference(interp::EnzymeInterpeter, mi::MethodInstance) = nothing
+Core.Compiler.unlock_mi_inference(interp::EnzymeInterpeter, mi::MethodInstance) = nothing
+
+function Core.Compiler.add_remark!(interp::EnzymeInterpeter, sv::InferenceState, msg)
+end
+
+Core.Compiler.may_optimize(interp::EnzymeInterpeter) = true
+Core.Compiler.may_compress(interp::EnzymeInterpeter) = true
+Core.Compiler.may_discard_trees(interp::EnzymeInterpeter) = true
+if VERSION >= v"1.7.0-DEV.577"
+Core.Compiler.verbose_stmt_info(interp::EnzymeInterpeter) = false
+end
+
+if isdefined(Base.Experimental, Symbol("@overlay"))
+Core.Compiler.method_table(interp::EnzymeInterpeter, sv::InferenceState) =
+    Core.Compiler.OverlayMethodTable(interp.world, interp.method_table)
+else
+Core.Compiler.method_table(interp::EnzymeInterpeter, sv::InferenceState) =
+    GPUCompiler.WorldOverlayMethodTable(interp.world)
+end
+
+function is_primitive_func(@nospecialize(TT))
+    if TT <: Tuple{typeof(log), Float64}
+        return true
+    end
+    return false
+end
+
+
+# branch on https://github.com/JuliaLang/julia/pull/41328
+@static if isdefined(Core.Compiler, :is_stmt_inline)
+# function Core.Compiler.inlining_policy(
+#     interp::EnzymeInterpeter, @nospecialize(src), stmt_flag::UInt8,
+#     mi::MethodInstance, argtypes::Vector{Any})
+#     @show typeof(src)
+#     if src === nothing
+#         if Compiler.is_stmt_inline(stmt_flag) || src.isinlineable
+#             return src.ir
+#         end
+#     else
+#         # the default inlining policy may try additional effor to find the source in a local cache
+#         return Base.@invoke Core.Compiler.inlining_policy(
+#             interp::AbstractInterpreter, nothing, stmt_flag::UInt8,
+#             mi::MethodInstance, argtypes::Vector{Any})
+#     end
+#     return nothing
+# end
+else # @static if isdefined(Compiler, :is_stmt_inline)
+
+import Core.Compiler: InliningTodo, InliningState
+enzyme_inlining_policy(src) = Core.Compiler.default_inlining_policy(src)
+Core.Compiler.inlining_policy(::EnzymeInterpeter) = enzyme_inlining_policy
+function Core.Compiler.resolve_todo(todo::InliningTodo, state::InliningState{S, T, <:typeof(enzyme_inlining_policy)}) where {S<:Union{Nothing, Core.Compiler.EdgeTracker}, T}
+    mi = todo.mi
+    @show mi.specTypes
+    @show is_primitive_func(mi.specTypes)
+    if is_primitive_func(mi.specTypes)
```

Comment:
`autodiff(foo, Active(2.0))` certainly hits this case.
It seems like you need to fix another part of Enzyme as well?
```julia
julia> autodiff(foo, Active(2.0)) # ERROR: LLVM error: unknown binary operator, see below for details
mi.specTypes = Tuple{typeof(log), Float64}
ERROR: MethodError: no method matching classify_arguments(::GPUCompiler.CompilerJob{Enzyme.Compiler.EnzymeTarget, Enzyme.Compiler.EnzymeCompilerParams, GPUCompiler.FunctionSpec{typeof(foo), Tuple{Float64}}}, ::LLVM.FunctionType)
Closest candidates are:
  classify_arguments(::GPUCompiler.CompilerJob, ::LLVM.Function) at ~/.julia/packages/GPUCompiler/fG3xK/src/irgen.jl:294
Stacktrace:
 [1] lower_convention(job::GPUCompiler.CompilerJob, mod::LLVM.Module, entry_f::LLVM.Function, actualRetType::Type)
   @ Enzyme.Compiler ~/julia/packages/Enzyme/src/compiler.jl:2007
 [2] codegen(output::Symbol, job::GPUCompiler.CompilerJob{Enzyme.Compiler.EnzymeTarget, Enzyme.Compiler.EnzymeCompilerParams, GPUCompiler.FunctionSpec{typeof(foo), Tuple{Float64}}}; libraries::Bool, deferred_codegen::Bool, optimize::Bool, strip::Bool, validate::Bool, only_entry::Bool, parent_job::Nothing)
   @ Enzyme.Compiler ~/julia/packages/Enzyme/src/compiler.jl:2273
 [3] _thunk(job::GPUCompiler.CompilerJob{Enzyme.Compiler.EnzymeTarget, Enzyme.Compiler.EnzymeCompilerParams, GPUCompiler.FunctionSpec{typeof(foo), Tuple{Float64}}})
   @ Enzyme.Compiler ~/julia/packages/Enzyme/src/compiler.jl:2628
 [4] cached_compilation(cache::Dict{UInt64, Any}, job::GPUCompiler.CompilerJob, compiler::typeof(Enzyme.Compiler._thunk), linker::typeof(Enzyme.Compiler._link))
   @ GPUCompiler ~/.julia/packages/GPUCompiler/fG3xK/src/cache.jl:89
 [5] thunk(f::typeof(foo), df::Nothing, ::Type{Active{Float64}}, tt::Type{Tuple{Active{Float64}}}, ::Val{false})
   @ Enzyme.Compiler ~/julia/packages/Enzyme/src/compiler.jl:2680
 [6] autodiff
   @ ~/julia/packages/Enzyme/src/Enzyme.jl:186 [inlined]
 [7] autodiff(f::typeof(foo), args::Active{Float64})
   @ Enzyme ~/julia/packages/Enzyme/src/Enzyme.jl:216
 [8] top-level scope
   @ REPL[3]:1
```

Reply:
Which version of Julia are you on? I am testing with 1.7 and you seem to have gotten an old version of GPUCompiler.

Reply:
You are right... It does work. Yay!

Reply:
Yeah, after updating deps, now it works:)